components/brave_wallet_ui/page/screens/composer_ui/select_token_modal/select_token_modal.tsx

```diff
@@ -167,270 +235,296 @@ export const SelectTokenModal = React.forwardRef<HTMLDivElement, Props>(
+              .length > 0
         )
+      }
 
-        if (
-          accountsLocalHost &&
-          account.accountId.keyringId !== BraveWallet.KeyringId.kFilecoin
-        ) {
-          list.push(accountsLocalHost)
-          return list
-        }
-
-        return list.filter((token) => token.visible)
-      },
-      [userVisibleTokensInfo, networks]
-    )
-
-    const getTokenListWithBalances = React.useCallback(
-      (account: BraveWallet.AccountInfo) => {
-        return getTokenListByAccount(account).filter((token) =>
-          new Amount(
-            getBalance(account.accountId, token, tokenBalancesRegistry)
-          ).gt(0)
-        )
-      },
-      [getTokenListByAccount, tokenBalancesRegistry]
-    )
+      return userVisibleFungibleTokens.filter(
+        (token) =>
+          getAllAccountsWithBalance(token, accounts, tokenBalancesRegistry)
+            .length > 0
+      )
+    }, [
+      modalType,
+      fullVisibleTokenList,
+      userVisibleNfts,
+      userVisibleFungibleTokens,
+      swapNetworksChainIds,
+      selectedSendOption,
+      accounts,
+      tokenBalancesRegistry
+    ])
 
-    const getTokensBySelectedSendOption = React.useCallback(
-      (account: BraveWallet.AccountInfo) => {
-        if (selectedSendOption === SendPageTabHashes.nft) {
-          return getTokenListWithBalances(account).filter(
-            (token) => token.isErc721 || token.isNft || token.isErc1155
-          )
+    const userTokenList: UserAssetInfoType[] = React.useMemo(() => {
+      if (!tokenBalancesRegistry) {
+        // wait for balances before computing this list
+        return []
+      }
+      return tokensBySelectedComposerOption.map((asset) => {
+        return {
+          asset,
+          assetBalance:
+            selectedAccountFilter.address === AllAccountsOption.address
+              ? getFullAssetBalance(
+                  asset,
+                  networks,
+                  accounts,
+                  tokenBalancesRegistry
+                )
+              : getBalance(
+                  selectedAccountFilter.accountId,
+                  asset,
+                  tokenBalancesRegistry
+                )
         }
-        return getTokenListWithBalances(account).filter(
-          (token) => !token.isErc721 && !token.isErc1155 && !token.isNft
-        )
-      },
-      [getTokenListWithBalances, selectedSendOption]
-    )
+      })
+    }, [
+      tokensBySelectedComposerOption,
+      tokenBalancesRegistry,
+      selectedAccountFilter,
+      accounts,
+      networks
+    ])
 
     const tokenPriceIds = React.useMemo(
       () =>
-        accounts
-          .flatMap(getTokensBySelectedSendOption)
+        userTokenList
           .filter(
-            (token) => !token.isErc721 && !token.isErc1155 && !token.isNft
+            (token) =>
+              !token.asset.isErc721 &&
+              !token.asset.isErc1155 &&
+              !token.asset.isNft
```

Comment:
perf: i believe that `isNft` works for everything now, so we should check that first (check others afterwards for a fail-safe)

Reply:
Moved!